### PR TITLE
G-API: cv::gapi::goodFeaturesToTrack() implementation

### PR DESCRIPTION
**Pull request by OrestChura**
### goodFeaturesToTrack() kernel is included in the standard CPU G-API kernel pack now
 - there are performance and accuracy tests
 - additional comparison functions and fixtures implemented

### Pull Request Readiness Checklist
- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is accuracy test, performance test _and test data in opencv_extra repository, if applicable_
      _Patch to opencv_extra has the same branch name._
- [x] The feature is well documented _and sample code can be built with the project CMake_

```
build_image:Custom=ubuntu-clang:18.04
build_gapi_standalone:Custom=ade-0.1.1f

force_builders=Custom,Custom Win,Custom Mac
Xbuild_image:Custom=ubuntu-openvino-2019r2.0:16.04
build_image:Custom Win=openvino-2019r2.0
build_image:Custom Mac=openvino-2019r2.0

test_modules:Custom=gapi
test_modules:Custom Win=gapi
test_modules:Custom Mac=gapi

buildworker:Custom=linux-1
test_opencl:Custom=ON
test_bigdata:Custom=1
test_filter:Custom=*

build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
Xbuild_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f
build_gapi_standalone:Custom Win=ade-0.1.1f

```

### CODE REVIEW

**Comment by OrestChura** [comment]
@dmatveev @rgarnov @AsyaPronina @smirnov-alexey please review

File: modules/gapi/include/opencv2/gapi/imgproc.hpp [hunk 1]
```diff
@@ -642,6 +650,53 @@ L2gradient=false ).
...
+CV_8UC1 and the same size as image ), it specifies the region in which the corners are detected.
+@param blockSize Size of an average block for computing a derivative covariation matrix over each
+pixel neighborhood. See cornerEigenValsAndVecs .
+@param useHarrisDetector Parameter indicating whether to use a Harris detector (see #cornerHarris)
+or #cornerMinEigenVal.
+@param k Free parameter of the Harris detector.
+ */
+GAPI_EXPORTS GArray<Point2f> goodFeaturesToTrack(const GMat& image, int maxCorners,
+                                                 double qualityLevel, double minDistance,
+                                                 const Mat& mask = Mat(), int blockSize = 3,
```
Comment (OrestChura) [inline modules/gapi/include/opencv2/gapi/imgproc.hpp]: Should the _mask_ parameter be Mat or GMat?
Or should it be removed at all? Maybe this functionality is unnecessary and should be cut off? Masks are ROIs most of the time so you can just put an appropriate ROI to the function instead of putting the whole image and a mask

File: modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp [hunk 2]
```diff
@@ -622,6 +622,54 @@ PERF_TEST_P_(CannyPerfTest, TestPerformance)
...
+    initMatrixRandN(type, sz, CV_8UC1, false);
+
+    std::vector<cv::Point2f> out_vec_ocv, out_vec_gapi;
+    out_vec_ocv.reserve(maxCorn);
+    out_vec_gapi.reserve(maxCorn);
+
+    // OpenCV code /////////////////////////////////////////////////////////////
+    {
+        cv::goodFeaturesToTrack(in_mat1, out_vec_ocv, maxCorn, qual, minDist, cv::noArray(), blockSize, useHarris, k);
+        out_mat_ocv = cv::Mat(out_vec_ocv);
```
Comment (OrestChura) [inline modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp]: Does it look OK to convert vector to Mat just not to implement compareVector() function?

File: modules/ts/include/opencv2/ts/ts_gtest.h [hunk 3]
```diff
@@ -17478,6 +17478,277 @@ class CartesianProductGenerator10
 };  // class CartesianProductGenerator10
 
 
+template <typename T1, typename T2, typename T3, typename T4, typename T5,
```
Comment (OrestChura) [inline modules/ts/include/opencv2/ts/ts_gtest.h]: I hope it's OK that I put additional classes in that file

Comment (TolyaTalamanov) [reply modules/ts/include/opencv2/ts/ts_gtest.h]: I think it's a bad idea :)

Comment (TolyaTalamanov) [reply modules/ts/include/opencv2/ts/ts_gtest.h]: Lets to discuss it locally

File: modules/gapi/include/opencv2/gapi/imgproc.hpp [hunk 4]
```diff
@@ -102,6 +104,12 @@ namespace imgproc {
         }
     };
 
+    G_TYPED_KERNEL(GGoodFeatures, <GArrayPoint2f(GMat,int,double,double,Mat,int,bool,double)>, "org.opencv.imgproc.goodFeatures") {
+        static GArrayDesc outMeta(GMatDesc, int, double, double, Mat, int, bool, double) {
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/imgproc.hpp]: Mat -> const cv::Mat&

File: modules/gapi/include/opencv2/gapi/imgproc.hpp (hunk 1 above)
Comment (dmatveev) [reply modules/gapi/include/opencv2/gapi/imgproc.hpp]: Let it be Mat. We need to preserve feature-parity with OpenCV when exposing OpenCV operations

File: modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp [hunk 5]
```diff
@@ -622,6 +622,54 @@ PERF_TEST_P_(CannyPerfTest, TestPerformance)
...
+    compare_f cmpF;
+    MatType type;
+    int maxCorn = 0, blockSize = 0;
+    double qual = 0.0, minDist = 0.0, k = 0.0;
+    cv::Size sz;
+    bool useHarris = false;
+    cv::GCompileArgs compile_args;
+    std::tie(cmpF, type, sz, maxCorn, qual, minDist, blockSize, useHarris, k, compile_args) = GetParam();
+
+    initMatrixRandN(type, sz, CV_8UC1, false);
```
Comment (dmatveev) [inline modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp]: Not sure if this is a correct way to test kernels like GoodFeatures.

Please find an appropriate image from `opencv_extra` and load it here. Baboon should work ok for the purpose:

image

File: modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp (hunk 2 above)
Comment (dmatveev) [reply modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp]: I think you need a custom comparison function here.

File: modules/gapi/perf/cpu/gapi_imgproc_perf_tests_cpu.cpp [hunk 6]
```diff
@@ -134,6 +134,18 @@ INSTANTIATE_TEST_CASE_P(CannyPerfTestCPU, CannyPerfTest,
...
+INSTANTIATE_TEST_CASE_P(GoodFeaturesPerfTestCPU, GoodFeaturesPerfTest,
+    Combine(Values(AbsExact().to_compare_f()),
+        Values(CV_8UC1),
+        Values(szVGA, sz720p, sz1080p),
+        Values(3, 0),
+        Values(0.5, 0.01),
+        Values(3.0, 30.0),
+        Values(3, 9),
+        Values(true, false),
+        Values(0.04, 0.5),
```
Comment (dmatveev) [inline modules/gapi/perf/cpu/gapi_imgproc_perf_tests_cpu.cpp]: Do we need so many values in the performance test?

File: modules/gapi/test/common/gapi_imgproc_tests.hpp [hunk 7]
```diff
@@ -52,6 +52,8 @@ GAPI_TEST_FIXTURE(SobelXYTest, initMatrixRandN, FIXTURE_API(CompareMats,int,int,
 GAPI_TEST_FIXTURE(EqHistTest, initMatrixRandN, FIXTURE_API(CompareMats), 1, cmpF)
 GAPI_TEST_FIXTURE(CannyTest, initMatrixRandN, FIXTURE_API(CompareMats,double,double,int,bool), 5,
     cmpF, thrLow, thrUp, apSize, l2gr)
+GAPI_TEST_FIXTURE(GoodFeaturesTest, initMatrixRandN, FIXTURE_API(CompareMats,int,double,double,int,bool,double), 7,
```
Comment (dmatveev) [inline modules/gapi/test/common/gapi_imgproc_tests.hpp]: Please use another input image to test a corner detector

File: modules/ts/include/opencv2/ts/ts_gtest.h (hunk 3 above)
Comment (dmatveev) [reply modules/ts/include/opencv2/ts/ts_gtest.h]: Instead you could just wrap parameters of your kernel to a structure and instantiate the test with two-three versions of this structure.

@alalek will also mention to replace this structure by enum, probably a good way to go as well.

File: modules/gapi/include/opencv2/gapi/imgproc.hpp (hunk 4 above)
Comment (OrestChura) [reply modules/gapi/include/opencv2/gapi/imgproc.hpp]: OK.
But want to draw attention that there are other kernels in the file (and I'm afraid not only in that one) where outmeta parameters are not const. Should I create a task on that issue?

Comment (dmatveev) [reply modules/gapi/include/opencv2/gapi/imgproc.hpp]: Yes please.

File: modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp (hunk 2 above)
Comment (OrestChura) [reply modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp]: implemented

File: modules/ts/include/opencv2/ts/ts_gtest.h (hunk 3 above)
Comment (OrestChura) [reply modules/ts/include/opencv2/ts/ts_gtest.h]: number of parameters was decreased according to the tests in imgproc

File: modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp (hunk 5 above)
Comment (OrestChura) [reply modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp]: done. Had to use initTestDataPath() function from Streaming test

**Comment by OrestChura** [comment]
@dmatveev @TolyaTalamanov pls review changes

File: modules/gapi/test/common/gapi_tests_common.hpp [hunk 8]
```diff
@@ -169,6 +185,10 @@ using compare_f = std::function<bool(const cv::Mat &a, const cv::Mat &b)>;
 
 using compare_scalar_f = std::function<bool(const cv::Scalar &a, const cv::Scalar &b)>;
 
+template<typename pointtype>
+using compare_vectorPoints_f = std::function<bool(const std::vector<cv::Point_<pointtype>> &a,
```
Comment (OrestChura) [inline modules/gapi/test/common/gapi_tests_common.hpp]: Does this template usage look good in this place? As optFlow kernel is WIP, I can say that vectors comparison function would be needed.
But this template goes like vector<Point_<T>> while optFlow would require vector<uchar> and vector<double> so the template should be vector<T>. Am I right?

File: modules/gapi/perf/cpu/gapi_imgproc_perf_tests_cpu.cpp (hunk 6 above)
Comment (OrestChura) [reply modules/gapi/perf/cpu/gapi_imgproc_perf_tests_cpu.cpp]: quantity decreased

File: modules/gapi/test/common/gapi_imgproc_tests.hpp (hunk 7 above)
Comment (OrestChura) [reply modules/gapi/test/common/gapi_imgproc_tests.hpp]: done

File: modules/gapi/test/common/gapi_tests_common.hpp (hunk 8 above)
Comment (OrestChura) [reply modules/gapi/test/common/gapi_tests_common.hpp]: comparison of vectors is unified to goodF and optFlow. Some posiible issues as wrong/inexact sorting of two compared vectors are not taken into account

**Comment by OrestChura** [comment]
@Volskig pls review

**Comment by dmatveev** [comment]
Be right back here after @Volskig and @TolyaTalamanov 's reviews :)

File: modules/gapi/src/backends/cpu/gcpuimgproc.cpp [hunk 9]
```diff
@@ -180,6 +180,14 @@ GAPI_OCV_KERNEL(GCPUCanny, cv::gapi::imgproc::GCanny)
     }
 };
 
+GAPI_OCV_KERNEL(GCPUGoodFeatures, cv::gapi::imgproc::GGoodFeatures)
+{
+    static void run(const cv::Mat& in, int maxCorners, double quality, double minDist, const cv::Mat& mask, int blockSize, bool useHarris, double k, std::vector<cv::Point2f> &out)
```
Comment (Volskig) [inline modules/gapi/src/backends/cpu/gcpuimgproc.cpp]: String too long, split it. Four-five parameters per line for example.

File: modules/gapi/test/common/gapi_imgproc_tests_inl.hpp [hunk 10]
```diff
@@ -380,6 +380,32 @@ TEST_P(CannyTest, AccuracyTest)
...
+    std::vector<cv::Point2f> out_vec_ocv, out_vec_gapi;
+
+    // G-API code //////////////////////////////////////////////////////////////
+    cv::GMat in;
+    auto out = cv::gapi::goodFeaturesToTrack(in, maxCorn, qual, minDist, cv::Mat(),
+                                             blockSize, useHarris, k);
+
+    cv::GComputation c(cv::GIn(in), cv::GOut(out));
+    c.apply(cv::gin(in_mat1), cv::gout(out_vec_gapi), getCompileArgs());
+    // OpenCV code /////////////////////////////////////////////////////////////
```
Comment (AsyaPronina) [inline modules/gapi/test/common/gapi_imgproc_tests_inl.hpp]: suggest to add a newline before

File: modules/gapi/test/common/gapi_tests_common.hpp [hunk 11]
```diff
@@ -188,7 +207,7 @@ struct MatType2
     int _value;
 };
 
-// Universal parameter wrapper for common (pre-defined) and specific (user-defined) parameters
+// Universal parameter wrappers for common (pre-defined) and specific (user-defined) parameters
```
Comment (AsyaPronina) [inline modules/gapi/test/common/gapi_tests_common.hpp]: Are you sure? I don't think that author means exactly such explanation, let's discuss

File: modules/gapi/test/common/gapi_tests_common.hpp [hunk 12]
```diff
@@ -272,6 +349,13 @@ struct TestWithParamBase : TestFunctional,
         Fixture() { InitF(type, sz, dtype); } \
     };
 
+#define GAPI_TEST_FIXTURE_NO_INIT(Fixture, API, Number, ...) \
```
Comment (AsyaPronina) [inline modules/gapi/test/common/gapi_tests_common.hpp]: Please replace GAPI_TEST_FIXTURE_NO_INIT macro and all code base developed to support it with the appropriate feature from another your [#16788](https://github.com/opencv/opencv/pull/16788/files) PR once you are done with one.

File: modules/gapi/src/backends/cpu/gcpuimgproc.cpp (hunk 9 above)
Comment (AsyaPronina) [reply modules/gapi/src/backends/cpu/gcpuimgproc.cpp]: Very very minor comment: suggest to preserve variables naming among the G-API operation wrapper, kernel's run method and every other place where you expect to face variables reflecting the same entities (ex: test parameters) to give the user opportunity to perceive the code in different places more easily.

**Comment by dmatveev** [comment]
Is it ready for review now?

File: modules/gapi/src/backends/cpu/gcpuimgproc.cpp (hunk 9 above)
Comment (OrestChura) [reply modules/gapi/src/backends/cpu/gcpuimgproc.cpp]: splitted; names preserved

File: modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp [hunk 13]
```diff
@@ -622,6 +622,52 @@ PERF_TEST_P_(CannyPerfTest, TestPerformance)
 
 //------------------------------------------------------------------------------
 
+PERF_TEST_P_(GoodFeaturesPerfTest, TestPerformance)
+{
+    initTestDataPath();
+    compare_vector_f<cv::Point2f> cmpF;
+    std::string fileName = "";
+    int maxCorn = 0, blockSize = 0;
```
[the remark this answers is not in the document]
Comment (OrestChura) [reply modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp]: it is for the KlockWork

File: modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp [hunk 14]
```diff
@@ -622,6 +622,52 @@ PERF_TEST_P_(CannyPerfTest, TestPerformance)
...
+    double qual = 0.0;
+    bool useHarris = false;
+    cv::GCompileArgs compile_args;
```
[the remark this answers is not in the document]
Comment (OrestChura) [reply modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp]: agree

File: modules/gapi/test/common/gapi_imgproc_tests_inl.hpp (hunk 10 above)
Comment (OrestChura) [reply modules/gapi/test/common/gapi_imgproc_tests_inl.hpp:399]: added

File: modules/gapi/test/common/gapi_tests_common.hpp (hunk 12 above)
Comment (OrestChura) [reply modules/gapi/test/common/gapi_tests_common.hpp]: done

File: modules/gapi/test/common/gapi_tests_common.hpp [hunk 15]
```diff
@@ -218,17 +266,24 @@ struct Params
...
-    using AllParams = Params<SpecificParams...>;
+    static constexpr const size_t compile_args_num = 3;
+};
 
-    MatType2 type = getCommonParam<0>();
-    cv::Size sz = getCommonParam<1>();
-    MatType2 dtype = getCommonParam<2>();
+template<typename ...SpecificParams>
+struct ParamsSpecific : public ParamsBase<std::tuple<cv::GCompileArgs(*)()>,
+                                               std::tuple<SpecificParams...>>
```
Comment (Volskig) [inline modules/gapi/test/common/gapi_tests_common.hpp:278]: Alignment. It may be better to put, as in [271](https://github.com/opencv/opencv/blob/8c0ae785562b7b52d1feba4fa2ed4e7f4d39b1ef/modules/gapi/test/common/gapi_tests_common.hpp) string.

File: modules/gapi/include/opencv2/gapi/imgproc.hpp [hunk 16]
```diff
@@ -102,6 +102,14 @@ namespace imgproc {
         }
     };
 
+    G_TYPED_KERNEL(GGoodFeatures, <cv::GArray<cv::Point2f>(GMat,int,double,double,
+                                                           Mat,int,bool,double)>,
+                   "org.opencv.imgproc.goodFeaturesToTrack") {
```
[the remark this answers is not in the document]
Comment (OrestChura) [reply modules/gapi/include/opencv2/gapi/imgproc.hpp:107]: No, I think it's not broken: the string is a parameter for `G_TYPED_KERNEL()` macro so it is aligned with it's parenthesis

File: modules/gapi/include/opencv2/gapi/imgproc.hpp [hunk 17]
```diff
@@ -657,6 +665,59 @@ L2gradient=false ).
...
+@param blockSize Size of an average block for computing a derivative covariation matrix over each
+pixel neighborhood. See cornerEigenValsAndVecs .
+@param useHarrisDetector Parameter indicating whether to use a Harris detector (see #cornerHarris)
+or #cornerMinEigenVal.
+@param k Free parameter of the Harris detector.
+
+@return vector of detected corners.
+ */
+GAPI_EXPORTS GArray<Point2f> goodFeaturesToTrack(const GMat  &image,
+                                                       int    maxCorners,
```
[the remark this answers is not in the document]
Comment (OrestChura) [reply modules/gapi/include/opencv2/gapi/imgproc.hpp:713]: Yeah, it seems ot be.. but I aligned the types and the references/places for references to see easily if the parameter is const reference

[the remark this answers is not in the document]
Comment (OrestChura) [reply modules/gapi/include/opencv2/gapi/imgproc.hpp:713]: but I see the broken alignment below. Thanks, will fix it!

File: modules/gapi/test/common/gapi_imgproc_tests.hpp [hunk 18]
```diff
@@ -52,6 +53,11 @@ GAPI_TEST_FIXTURE(SobelXYTest, initMatrixRandN, FIXTURE_API(CompareMats,int,int,
 GAPI_TEST_FIXTURE(EqHistTest, initMatrixRandN, FIXTURE_API(CompareMats), 1, cmpF)
 GAPI_TEST_FIXTURE(CannyTest, initMatrixRandN, FIXTURE_API(CompareMats,double,double,int,bool), 5,
     cmpF, thrLow, thrUp, apSize, l2gr)
+GAPI_TEST_FIXTURE_SPEC_PARAMS(GoodFeaturesTest, FIXTURE_API(CompareVectors<cv::Point2f>,
```
[the remark this answers is not in the document]
Comment (OrestChura) [reply modules/gapi/test/common/gapi_imgproc_tests.hpp]: I made little changes. Looks better now?

File: modules/gapi/test/common/gapi_tests_common.hpp (hunk 15 above)
Comment (OrestChura) [reply modules/gapi/test/common/gapi_tests_common.hpp]: agree, fixed

File: modules/gapi/include/opencv2/gapi/imgproc.hpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/include/opencv2/gapi/imgproc.hpp:107]: Ok

File: modules/gapi/include/opencv2/gapi/imgproc.hpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/include/opencv2/gapi/imgproc.hpp:713]: Ok

File: modules/gapi/test/common/gapi_imgproc_tests.hpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/common/gapi_imgproc_tests.hpp]: Yes

**Comment by alalek** [comment]
> squash! Full goodFeaturesToTrack implementations and tests …

Don't use such prefixes in commit messages.
Prefer using "gapi: " prefix for commits with Graph API module changes.

**Comment by OrestChura** [comment]
OK, sorry. Thank you!
